components/core/src/clp/StreamingReader.hpp

```diff
@@ -0,0 +1,339 @@
+#include <string_view>
+#include <thread>
+#include <vector>
+
+#include <curl/curl.h>
+
+#include "Defs.h"
+#include "ErrorCode.hpp"
+#include "ReaderInterface.hpp"
+#include "Thread.hpp"
+#include "TraceableException.hpp"
+
+namespace clp {
+/**
+ * This class implements a ReaderInterface to stream data directly from a given URL (e.g., streaming
+ * from a pre-signed S3 URL). When opening a URL, a thread will be created to transfer data in
+ * the background using libcurl.
+ *
+ * This class guarantees synchronization between the read operations and data streaming
+ * (downloading). The reader is only designed for single threaded sequencial read.
+ */
+class StreamingReader : public ReaderInterface {
+public:
+    using BufferView = std::span<char>;
+
+    // Types
+    /**
+     * The exception thrown by this class.
+     */
+    class OperationFailed : public TraceableException {
+    public:
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return "clp::StreamingReader operation failed.";
+        }
+    };
+
+    /**
+     * Enum defining the states of the reader.
+     */
+    enum class State : uint8_t {
+        // InProgress: The reader is transferring data from the source URL.
+        InProgress = 0,
+        // Failed: The reader has failed to stream data from the source URL.
+        Failed,
+        // Finished: The streaming of the given source URL has been accomplished.
+        Finished
+    };
+
+    static constexpr size_t cDefaultBufferPoolSize{8};
+    static constexpr size_t cDefaultBufferSize{4096};
+
+    static constexpr size_t cMinBufferPoolSize{2};
+    static constexpr size_t cMinBufferSize{512};
+
+    /**
+     * These static members defines the default connection timeout and operation timeout.
+     * Please refer to the libcurl documentation for more details:
+     * ConnectionTimeout: https://curl.se/libcurl/c/CURLOPT_CONNECTTIMEOUT.html
+     * OverallTimeout: https://curl.se/libcurl/c/CURLOPT_TIMEOUT.html
+     */
+    static constexpr uint32_t cDefaultConnectionTimeout{10};  // Seconds
+    static constexpr uint32_t cDefaultOverallTimeout{45};  // Seconds
+
+    /**
+     * Initializes the underlying libcurl functionalities globally. It should be called before
+     * opening a source URL for data transferring.
+     * @return ErrorCode_Success on success.
+     * @return ErrorCode_Failure if libcurl initialization failed.
+     */
+    [[nodiscard]] static auto init() -> ErrorCode;
+
+    /**
+     * Releases the globally initialized libcurl resources.
+     */
+    static auto deinit() -> void;
+
+    /**
+     * Constructs a new connection to stream data from the given url with a given offset.
+     * TODO: the current implementation doesn't handle the case when the given offset is out of
```

Comment:
Sorry, forgot to mention this in the previous review:

* Conceptually, we should be setting `m_file_pos` to the offset requested by the caller, right?
* Let's say the user chooses an offset that's out of range, wouldn't the download fail and so they wouldn't be able to read anything anyway?

Reply:
- yes, we should set `m_file_pos`
- the download won't fail. It will exit peacefully and return `CURLE_OK` even thought the offset is out of bound

Reply:
Update: we can sest the handle to fail when HTTP fails